dipy/tracking/streamline.py

```diff
@@ -323,3 +327,138 @@ def orient_by_rois(streamlines, roi1, roi2, affine=None, copy=True):
             new_sl[idx] = sl[::-1]
 
     return new_sl
+
+
+def _extract_vals(data, streamlines, affine=None, threedvec=False):
+    """
+    Helper function for use with `values_from_volume`.
+
+    Parameters
+    ----------
+    data : 3D or 4D array
+        Scalar (for 3D) and vector (for 4D) values to be extracted. For 4D
+        data, interpolation will be done on the 3 spatial dimensions in each
+        volume.
+
+    streamlines : ndarray or list
+        If array, of shape (n_streamlines, n_nodes, 3)
+        If list, len(n_streamlines) with (n_nodes, 3) array in
+        each element of the list.
+
+    affine : ndarray, shape (4, 4)
+        Affine transformation from voxels (image coordinates) to streamlines.
+        Default: identity.
+
+    threedvec : bool
+        Whether the last dimension has length 3. This is a special case in
+        which we can use :func:`vfu.interpolate_vector_3d` for the
+        interploation of 4D volumes without looping over the elements of the
+        last dimension.
+
+    Return
+    ------
+    array or list (depending on the input) : values interpolate to each
+        coordinate along the length of each streamline
+    """
+    data = data.astype(np.float)
+    if (isinstance(streamlines, list) or
+            isinstance(streamlines, types.GeneratorType)):
+        if affine is not None:
+            streamlines = ut.move_streamlines(streamlines,
+                                              np.linalg.inv(affine))
+
+        vals = []
+        for sl in streamlines:
+            if threedvec:
+                vals.append(list(vfu.interpolate_vector_3d(data,
+                                 sl.astype(np.float))[0]))
+            else:
+                vals.append(list(vfu.interpolate_scalar_3d(data,
+                                 sl.astype(np.float))[0]))
+
+    elif isinstance(streamlines, np.ndarray):
+        sl_shape = streamlines.shape
+        sl_cat = streamlines.reshape(sl_shape[0] *
+                                     sl_shape[1], 3).astype(np.float)
+
+        if affine is not None:
+            inv_affine = np.linalg.inv(affine)
+            sl_cat = (np.dot(sl_cat, inv_affine[:3, :3]) +
```

Comment:
Yeah - but then you have to augment the coordinates, adding a '1' to every
coordinate. This is a way to get around that.


On Tue, Jan 26, 2016 at 6:28 AM, Samuel St-Jean <notifications@github.com>
wrote:

> In dipy/tracking/streamline.py
> <https://github.com/nipy/dipy/pull/811#discussion_r50840662>:
>
> > +        for sl in streamlines:
> > +            if threedvec:
> > +                vals.append(list(vfu.interpolate_vector_3d(data,
> > +                                 sl.astype(np.float))[0]))
> > +            else:
> > +                vals.append(list(vfu.interpolate_scalar_3d(data,
> > +                                 sl.astype(np.float))[0]))
> > +
> > +    elif isinstance(streamlines, np.ndarray):
> > +        sl_shape = streamlines.shape
> > +        sl_cat = streamlines.reshape(sl_shape[0] *
> > +                                     sl_shape[1], 3).astype(np.float)
> > +
> > +        if affine is not None:
> > +            inv_affine = np.linalg.inv(affine)
> > +            sl_cat = (np.dot(sl_cat, inv_affine[:3, :3]) +
>
> You can also do that in only one matrix multiplication, that's why
> homogeneous coordinates are used.
>
> https://en.wikipedia.org/wiki/Transformation_matrix#Affine_transformations
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/811/files#r50840662>.
>